TESTING.asciidoc

```diff
@@ -76,7 +76,177 @@ In order to start a node with a trial license execute the following command:
+
+Note that this will also run the unit tests and precommit tasks first. If you want to just
+run the integration tests (because you are debugging them):
+
+---------------------------------------------------------------------------
+./gradlew integTest
+---------------------------------------------------------------------------
+
+If you want to just run the precommit checks:
+
+---------------------------------------------------------------------------
+./gradlew precommit
+---------------------------------------------------------------------------
+
+Some of these checks will require `docker-compose` installed for bringing up
+test fixtures. If it's not present those checks will be skipped automatically.
+
+== How to write good tests?
+
+=== Test pyramid
+
+-------------------------------------------------
+                  /\
+                 /  \
+                /    \
+               /manual\
+              /________\ 
+             /          \
+            / 3rd party  \ 
+           /      OS      \   
+          /      BWC       \   
+         / Internal Cluster \ 
+        /  External (REST)   \ 
+       /        Unit          \
+      /________________________\
+-------------------------------------------------
+
+As a rule of thumb, always prefer writing tests at the lowest possible level 
+of the pyramid. 
+
+==== Unit Tests 
+
+Unit tests are the preferred way to test some functionality: most of the time
+they are simpler to understand, more likely to reproduce, and unlikely to be
+affected by changes that are unrelated to the piece of functionality that is
+being tested.
+
+**`ESTestCase`**: The base class of all tests. 
+It is typically extended directly by unit tests.
+
+The Gradle target to run unit tests is **`test`**.
+
+==== External REST Tests 
+
+The `'elasticsearch.testclusters'` Gradle plugin has support for configuring
+and starting, and stopping clusters used for testing. 
+These test run against a cluster started from the generated distribution with 
+a configuration specific for the test and only interact using the REST APIs. 
+
+These can use the full default or oss distribution but we can also use a special
+one (integ-test) built for testing only that doesn't have any modules added.
+Module that are needed by the test can be added trough the testclusters plugin.
+This approach is preferred as it leads to better isolation, faster execution,
+and better build avoidance: if a change only changes a particular module, 
+all the others don't have to be re-tested, except when testing using the full 
+distribution.
+
+**`ESRestTestCase`**: is the base class for these tests.
+These tests can also be implemented using YAML and ran using a subclass of
+**ESClientYamlSuiteTestCase**.
+
+The Gradle target to run rest tests is **`integTest`**
+
+==== Internal Cluster Tests
+
+There are multiple base classes for these:
+
+* **`ESSingleNodeTestCase`**: This test case sets up a cluster that has a
```

Comment:
ESSingleNodeTestCase are considered unit tests. AFAIK they are not run with internalClusterTest